### ISSUE TITLE
Inplace operators create an alias

### ISSUE THREAD

**Issue by NiklasGustafsson** [dotnet/TorchSharp#1064]
Several, if not all, in-place operators create an alias to the input tensor. While the operator does not allocate new tensor storage, it does allocate a new managed-code instance and increments the reference count on the native storage.

This is arguably not very intuitive. To address it would be a (mechanically speaking) rather large change to make, and a breaking change.

Thoughts -- @lostmsu, @kaiidams, @MovGP0, @dayo05, @ChengYen-Tang?

**Comment by ChengYen-Tang** [comment]
> Several, if not all, in-place operators create an alias to the input tensor. While the operator does not allocate new tensor storage, it does allocate a new managed-code instance and increments the reference count on the native storage.

https://github.com/dotnet/TorchSharp/pull/1053#issuecomment-1673929247
Sorry, I don't really understand what it means

**Comment by NiklasGustafsson** [comment]
> [#1053 (comment)](https://github.com/dotnet/TorchSharp/pull/1053#issuecomment-1673929247) Sorry, I don't really understand what it means

The data of tensors are held in native storage, and the (small) managed object representing a tensor has a handle (pointer) to the native storage. For most operations, the operation will create a new native object, and a new managed object. There's also `.alias()`, which creates a second managed object pointing to the same native object as the input.

It turns out that all (or most) of the in-place operators (ending with '_'), which modify the native object rather than creating a new one -- much more efficient, obviously -- return a second managed object, which will keep the reference-counted native object alive.

So, if you do something like this:

```C#
using var x = input.mul(5).add_(10);
```

The tensor returned by `mul(5)` will be modified in-place, correctly, but there will be two managed objects -- the one returned by `mul(5)` and the one returned by `add_(10)`. The latter is stored in 'x' and then disposed later. The former, however, is not disposed, and keeps the native storage from being reclaimed. If you have a dynamic dispose scope active, then it will be dealt with properly when the dispose scope is disposed, but not until then.

**Comment by NiklasGustafsson** [comment]
> > [#1053 (comment)](https://github.com/dotnet/TorchSharp/pull/1053#issuecomment-1673929247) Sorry, I don't really understand what it means
> 
> The data of tensors are held in native storage, and the (small) managed object representing a tensor has a handle (pointer) to the native storage. For most operations, the operation will create a new native object, and a new managed object. There's also `.alias()`, which creates a second managed object pointing to the same native object as the input.
> 
> It turns out that all (or most) of the in-place operators (ending with '_'), which modify the native object rather than creating a new one -- much more efficient, obviously -- return a second managed object, which will keep the reference-counted native object alive.
> 
> So, if you do something like this:
> 
> ```cs
> using var x = input.mul(5).add_(10);
> ```
> 
> The tensor returned by `mul(5)` will be modified in-place, correctly, but there will be two managed objects -- the one returned by `mul(5)` and the one returned by `add_(10)`. The latter is stored in 'x' and then disposed later. The former, however, is not disposed, and keeps the native storage from being reclaimed. If you have a dynamic dispose scope active, then it will be dealt with properly when the dispose scope is disposed, but not until then.

So, the argument is that this is non-intuitive, that if you use an inplace operator, you would expect it not to return a new managed tensor object (alias).

**Comment by ChengYen-Tang** [comment]
My current doubts:
1. Is there a way to delay execution like linq?
2. WeakReference, but Tensor exists Dispose.

**Comment by kaiidams** [comment]
It sounds correct to me that inline operators return the same managed object like PyTorch does. It make a breaking changes but I'm not sure how big impact it has. Normally we don't depends on the behavior where inline operators make another manged objects. Is there major usecases which it matters? 

```py
import torch
x = torch.zeros(100)
y = x.add_(100)
id(x), id(y)
(140694535087664, 140694535087664)
```

**Comment by dayo05** [comment]
> It sounds correct to me that inline operators return the same managed object like PyTorch does. It make a breaking changes but I'm not sure how big impact it has. Normally we don't depends on the behavior where inline operators make another manged objects. Is there major usecases which it matters?
> 
> ```python
> import torch
> x = torch.zeros(100)
> y = x.add_(100)
> id(x), id(y)
> (140694535087664, 140694535087664)
> ```

I agree with this too. This will give a huge impact which depends on default `Equals` method which rarely used on. Also, this library used on ML; usually highly isolated with other parts of software.

Also, I think almost of user claims to check entire tensor is same for checking equation, not the reference.

Is there any reasons like memory leaking on this? If not, It will be good to follow the PyTorch's behaviour but I think this should be the lower priority.
